Book/php7/internal_types/functions/callables.rst

```diff
@@ -83,21 +83,51 @@ Let detail the various FCC fields:
-  function.
-
 .. warning:: Prior to PHP 7.3.0 there existed an ``initialized`` field. Now an FCC is considered initialized when
   ``function_handler`` is set to a non-null pointer.
 
+The *only* case where an FCC will be uninitialized is if the function is a trampoline, i.e. when the method
+of a class does not exist but is handled by the magic methods ``__call()``/``__callStatic()``.
+This is because a trampoline is freed by ZPP as it is a newly allocated ``zend_function`` struct with the
+op array copied, and is freed when called. To retrieve it manually use ``zend_is_callable_ex()``.
+
+.. warning:: It is not sufficient to just store the FCC to be able to call a user function at a later stage.
+   If the callable zval from the FCI is an object (because it has an ``__invoke`` method, is a ``Closure``,
+   or a trampoline) then a reference to the ``zend_object`` must also be stored, the refcount incremented,
+   and released as needed. Moreover, if the callable is a trampoline the ``function_handler`` must be copied
+   to be persisted between calls (see how SPL implements the storage of autoloading functions).
+
+.. note::  To release a FCC you should use the ``void zend_release_fcall_info_cache(zend_fcall_info_cache *fcc)``
+  function if maintain copies of a potential trampoline, as this will release the trampoline properly.
+  Moreover, this needs to be called *prior* to freeing the closure, as the trampoline will partially refer to a
+  ``zend_function *`` entry in the closure CE.
+
+..
+    This API is still being worked on and won't be usable for a year
+    note:: As of PHP 8.3.0, the FCC holds a ``closure`` field and a dedicated API to handle storing userland callables.
+
 Zend Engine API for callables
 -----------------------------
 
-The API can be found in the ``Zend_API.h`` header file.
+The API is located at various locations in the ``Zend_API.h`` header file.
+We will describe the various APIs needed to deal with callables in PHP.
+
+First of all, to check if an FCI is initialized use the ``ZEND_FCI_INITIALIZED(fci)`` macro.
+
+.. And, as of PHP 8.3.0, the ``ZEND_FCC_INITIALIZED(fcc)`` macro to check if an FCC is initialized.
 
-If you have a FCI/FCC pair for a callable you can call it directly by using the
+If you have a correctly initialized and set up FCI/FCC pair for a callable you can call it directly by using the
 ``zend_call_function(zend_fcall_info *fci, zend_fcall_info_cache *fci_cache)`` function.
-If you just need to change, or provide the arguments and return value you can use the
-``zend_fcall_info_call(zend_fcall_info *fci, zend_fcall_info_cache *fcc, zval *retval, zval *args)`` function.
+
+.. warning:: The ``zend_fcall_info_arg*()`` and ``zend_fcall_info_call()`` APIs should not be used.
+   The reasons for this is because the ``zval *args`` parameter does *not* set the ``params`` field of the FCI,
+   but is expected to be a PHP array containing positional arguments. If this is the case the ``named_params``
```

Comment:
So `args` is expected to hold positional arguments? If so, the following sentence doesn't make sense. Should that be:
```suggestion
   but is expected to be a PHP array containing positional arguments. If this is not the case the ``named_params``
```

Reply:
> So `args` is expected to hold positional arguments? If so, the following sentence doesn't make sense. Should that be:

``args`` must be an IS_ARRAY zval which holds positional arguments, this can be seen looking at the implementation of [zend_fcall_info_args_ex()](https://heap.space/xref/php-src/Zend/zend_API.c?r=182314c3#zend_fcall_info_args_ex)
which will reallocate the entries in the array to push them to a C array of params.

The point I'm making is that ``named_params`` is a ``HashTable *``, that *also* accepts positional arguments, and that instead of using the API to, in most cases, needlessly reallocate the arguments to just provide the pointer of the HashTable to that field in the FCI.

Reply:
Pushed a commit to hopefully clarify what I mean.